prow/plank/reconciler.go

```diff
@@ -111,15 +105,49 @@ func add(
 	return nil
 }
 
+func newReconciler(pjClient ctrlruntimeclient.Client, overwriteReconcile reconcileFunc, cfg config.Getter, totURL string) *reconciler {
+	return &reconciler{
+		ctx:                context.Background(),
+		pjClient:           pjClient,
+		buildClients:       map[string]ctrlruntimeclient.Client{},
+		overwriteReconcile: overwriteReconcile,
+		log:                logrus.NewEntry(logrus.StandardLogger()).WithField("controller", ControllerName),
+		config:             cfg,
+		totURL:             totURL,
+		clock:              clock.RealClock{},
+		serializationLocks: &shardedLock{
+			mapLock: &sync.Mutex{},
+			locks:   map[string]*semaphore.Weighted{},
+		},
+	}
+}
+
+type reconcileFunc = func(reconcile.Request) (reconcile.Result, error)
+
 type reconciler struct {
 	ctx                context.Context
 	pjClient           ctrlruntimeclient.Client
 	buildClients       map[string]ctrlruntimeclient.Client
-	overwriteReconcile func(reconcile.Request) (reconcile.Result, error)
+	overwriteReconcile reconcileFunc
 	log                *logrus.Entry
 	config             config.Getter
 	totURL             string
 	clock              clock.Clock
+	serializationLocks *shardedLock
+}
+
+type shardedLock struct {
+	mapLock *sync.Mutex
+	locks   map[string]*semaphore.Weighted
+}
+
+func (s *shardedLock) getLock(key string) *semaphore.Weighted {
+	s.mapLock.Lock()
+	defer s.mapLock.Unlock()
+	if _, exists := s.locks[key]; !exists {
+		s.locks[key] = semaphore.NewWeighted(1)
```

Comment:
Why semaphore if you only have weight `1`?

Reply:
Because of the `TryAcquire`, `sync.Mutex` doesn't have something like that